Nojdi2

fundies opened the pull request.

fundies committed:
remerge

fundies committed:
all tests pass sometimes

fundies committed:
all games listed in pr build

JoshDreamland commented on CompilerSource/compiler/components/parse_and_link.cpp:
```diff
@@ -258,11 +260,33 @@ int lang_CPP::compile_parseAndLink(const GameData &game, CompileState &state) {
+          // Format DebugString output to replace escape sequences with newlines
+          std::string debug_str = event.DebugString();
+          std::string formatted_debug;
+          for (size_t i = 0; i < debug_str.length(); ++i) {
```
This probably doesn't belong here

fundies committed:
cmake fix

JoshDreamland committed:
Add lookahead for `TT_DOT` vs `TT_DECLITERAL` lexing

Actual implementation: check if the digit after '.' is numeric; if it is, reuse number parsing code in the switch statement; if not, jump directly into the '.' node of the token Trie.

fundies committed:
Fix variable extraction, std::string detection, DEBUG_MODE handling, and error reporting

Major fixes:

1. Variable extraction and scope handling
   - Filter out function-local variable declarations in clang adapter
     * Prevent CXCursor_VarDecl with function parents from being added to namespace/class scopes
     * Fixes issue where loop variables like 'i' were incorrectly added to enigma_user namespace
   - Prevent shared locals from being added to dot_accessed_locals
     * Check is_shared_local() in AddDot before adding to dot_accessed_locals
     * Fixes "no member named 'varaccess_x'" errors for shared locals like x, y
   - Generate direct member access for shared locals in pretty printer
     * Use direct member access instead of varaccess_ functions for shared locals

2. std::string detection and namespace handling
   - Add CXCursor_TypeAliasDecl handling in clang adapter
     * Recognize type aliases and assign DEF_TYPENAME | DEF_TYPED flags
     * Ensure type aliases have non-empty names before adding to scope
   - Implement using declaration processing
     * Add third pass in build_definitions() to process CXCursor_UsingDeclaration
     * Handle using std::string and using namespace declarations
     * Improve qualified name resolution using clang_getCursorUSR
   - Add inline namespace handling
     * Detect inline namespaces and propagate members to parent scope
     * Makes std::string (from std::__1) available directly in std namespace
   - Remove __1-specific fallback logic in favor of generic namespace handling

3. DEBUG_MODE conditional compilation
   - Make DEBUG_MODE define conditional on run mode
     * Only define DEBUG_MODE during parsing when mode is "Debug" (from -m flag)
     * Pass mode from emake to definitionsModified via YAML settings
     * Add target-mode field to APIyaml output
     * Parse mode from YAML in definitionsModified
   - debug_scope code generation already conditional on mode == emode_debug
     * Verified in write_object_data.cpp and write_room_data.cpp

4. Error handling and reporting
   - Make parse failures fatal
     * Set err_str via ide_passback_error.set() when parse_file fails
     * Ensures emake properly detects and reports parse errors
   - Fix definitionsModified return value
     * Return nullptr on success instead of &ide_passback_error
     * Refine emake error check to require both def_err and err_str
   - Fix missing space in success message

5. Platform-specific fixes
   - Add macOS Homebrew include path
     * Add /opt/homebrew/include/ to include search paths for macOS
     * Moved to clang_adapter.cpp with proper -I flag

6. Code cleanup
   - Remove all debug prints throughout codebase
     * Removed from collect_variables.cpp, jdi_utility.cpp, clang_adapter.cpp
     * Removed from lang_CPP.cpp, write_object_data.cpp, pretty_printer.cpp
     * Removed from parser.cpp, event_parser.cpp, proto_util.h
     * Removed object traversal debug prints

Files changed:
- CompilerSource/languages/clang_adapter.cpp:
  * Filter function-local variables in process_cursor()
  * Add CXCursor_TypeAliasDecl handling
  * Add inline namespace detection and member propagation
  * Implement third pass for using declarations
  * Add macOS Homebrew include path
  * Remove debug prints

- CompilerSource/parser/collect_variables.cpp:
  * Prevent shared locals from being added to dot_accessed_locals in AddDot()
  * Remove debug prints and workaround logic

- CompilerSource/parsing/pretty_printer.cpp:
  * Generate direct member access for shared locals in VisitDot()
  * Remove debug prints

- CompilerSource/languages/lang_CPP.cpp:
  * Conditional DEBUG_MODE based on mode from YAML
  * Set err_str on parse failure
  * Return nullptr on success
  * Fix missing space in success message
  * Remove debug prints

- CommandLine/emake/Main.cpp:
  * Get mode early and pass to APIyaml()
  * Refine error check to require err_str

- CommandLine/emake/OptionsParser.cpp/hpp:
  * Add mode parameter to APIyaml()
  * Add target-mode to YAML output

- CommandLine/emake/Server.cpp:
  * Pass default mode to APIyaml()

- CompilerSource/compiler/jdi_utility.cpp:
  * Remove debug prints

- CompilerSource/parsing/parser.cpp:
  * Remove debug prints

- CompilerSource/compiler/components/write_object_data.cpp:
  * Remove debug prints

- shared/event_reader/event_parser.cpp:
  * Remove debug prints

- shared/proto_util.h:
  * Remove debug prints

fundies committed:
Fix compiler warnings and lexer backslash handling

- Add override keywords to virtual function overrides in codegen
  * Add override to base class functions (gamestart, roomstart, draw, etc.) in event_parent
  * Add override to object class functions that override base functions
  * Only add override to actual overrides, not to new functions like alarm_0, collision_obj_fin

- Suppress -Wmissing-declarations for empty enums in resource names
  * Add pragma directives around enum generation in write_resource_meta

- Fix lexer handling of backslashes in macro translations
  * Handle unknown escape sequences gracefully (treat as literal backslash + char)
  * Handle line continuations (backslash + newline) in macros
  * Skip standalone backslashes silently to prevent "Unexpected symbol '\'" errors
  * Fixes errors when translating macros to EDL with Windows paths or other backslashes

fundies committed:
Remove all JUST_DEFINE_IT_RUN conditional code since clang_adapter
replaced JDI and no longer needs these workarounds. The macro is never
defined, so all #ifdef blocks were dead code and all #ifndef blocks
were always active.

- Removed dead code blocks from var4.h, graphics_object.h,
  collisions_object.h, mathnc.h, GTest/include.h, and GAME_GLOBALS.h
- Made all conditional code unconditional across the codebase
- Removed error checks that are no longer needed

fundies committed:
Fix parsing and codegen issues for GameMaker syntax compatibility

This commit addresses several parsing and code generation issues to improve
GameMaker syntax compatibility:

1. Fix repeat macro parsing:
   - Detect repeat(expr) form and parse it as a for loop internally while
     preserving the macro form in pretty printer output for readability
   - Prevent repeat from being used in function call expressions (it's a
     statement, not an expression)
   - Add test to verify repeat(expr) parsing works correctly

2. Fix variable scoping:
   - Variables declared in for loops and if statements (like ENIGMA_REPEAT_VAR)
     are now correctly identified as block-scoped and not written to object
     variables
   - Filter ENIGMA_REPEAT_VAR from object variable declarations

3. Fix keyword operator spacing:
   - Add spaces after keyword operators (not, and, or) in pretty printer
     output to prevent them from being concatenated with identifiers

4. Fix argument[] array syntax:
   - Declare variant argument[16] array in script functions, initialized
     with argument0 through argument15 parameters, enabling argument[index]
     syntax

5. Fix = to == conversion in conditionals:
   - Convert assignment operators to equality operators in if/while/else if
     statements via SyntaxChecker visitor
   - Properly track conditional context to apply transformation recursively

6. Fix string literal quote style:
   - Always emit double quotes for string literals (GameMaker convention)
     instead of single quotes for character literals

Changes:
- parser.cpp: Parse repeat(expr) as for loop, add syntax checker for = to ==
- pretty_printer.cpp: Detect repeat macro pattern, add keyword spacing, use double quotes
- collect_variables.cpp: Track block-scoped variables in for/if statements
- write_object_data.cpp: Declare argument array, filter ENIGMA_REPEAT_VAR
- parser-test.cpp: Add test for repeat macro parsing
- lexer-test.cpp, pretty-printer_test.cpp: Add tests for keyword spacing and quotes

fundies committed:
commit state

fundies committed:
tests passing, mario working

fundies committed:
save state

fundies committed:
push state

fundies committed:
more fixes

JoshDreamland committed:
Fix missing header from OS X

fundies committed:
add vscode settings
